src/Functions/array/arrayIndex.h

```diff
@@ -55,9 +68,23 @@ struct ArrayIndexNumImpl
 #pragma GCC diagnostic ignored "-Wsign-compare"
 
     /// compares `lhs` against `i`-th element of `rhs`
-    static bool compare(const T & lhs, const PaddedPODArray<U> & rhs, const size_t i) { return lhs == rhs[i]; }
+    static bool compare(const Initial & lhs, const PaddedPODArray<Result> & rhs, const size_t i) { return lhs == rhs[i]; }
+
     /// compares `lhs against `rhs`, third argument unused
-    static bool compare(const T & lhs, const U & rhs, size_t) { return lhs == rhs; }
+    static bool compare(const Initial & lhs, const Result & rhs, size_t) { return lhs == rhs; }
+
+    static Initial extract(const PaddedPODArray<Initial> & a, size_t i) { return a[i]; }
+
+    static Initial extract(const IColumn & a, size_t i)
+    {
+        if constexpr (std::is_same_v<Initial, Float32>)
+            return a.getFloat32(i);
+        if constexpr (std::is_same_v<Initial, Float64>)
+            return a.getFloat64(i);
+        if constexpr (is_signed_v<Initial>)
+            return a.getInt(i);
+        return a.getUInt(i);
```

Comment:
Please make sure that there is no way to instantiate `extract` with `Initial` = `String`

Reply:
It's the *NumImpl, so it's a pre-condition to invoke it with integral `Initial` and `Result` template parameters only.